r/sagemaker_code_repository - new resource

DrFaust92 opened the pull request:
<!--- See what makes a good Pull Request at : https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/CONTRIBUTING.md --->

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

<!--- Thank you for keeping this note for the community --->

<!--- If your PR fully resolves and should automatically close the linked issue, use Closes. Otherwise, use Relates --->
Closes #13817

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):
<!--
If change is not user facing, just write "NONE" in the release-note block below.
-->

```release-note
resource_aws_sagemaker_code_repository - new resource
```

Output from acceptance testing:

<!--
Replace TestAccXXX with a pattern that matches the tests affected by this PR.

For more information on the `-run` flag, see the `go test` documentation at https://tip.golang.org/cmd/go/#hdr-Testing_flags.
-->
```
$ make testacc TESTARGS='-run=TestAccAWSSagemakerCodeRepository_'
--- PASS: TestAccAWSSagemakerCodeRepository_disappears (26.45s)
--- PASS: TestAccAWSSagemakerCodeRepository_gitConfig_branch (35.76s)
--- PASS: TestAccAWSSagemakerCodeRepository_basic (36.04s)
--- PASS: TestAccAWSSagemakerCodeRepository_gitConfig_secret (84.66s)
```


`--- PASS: TestAccAWSSagemakerNotebookInstance_default_code_repository_sagemakerRepo (1165.81s)`

ewbankkit commented on aws/resource_aws_sagemaker_code_repository.go:
```diff
@@ -0,0 +1,205 @@
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringLenBetween(1, 63),
```
Add additional validation:

> [Valid characters are a-z, A-Z, 0-9, and - (hyphen).](https://docs.aws.amazon.com/sagemaker/latest/APIReference/API_CreateCodeRepository.html#API_CreateCodeRepository_RequestSyntax)

DrFaust92 replied:
done

ewbankkit commented on aws/resource_aws_sagemaker_code_repository.go:
```diff
@@ -0,0 +1,205 @@
+func resourceAwsSagemakerCodeRepositoryRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).sagemakerconn
+
+	describeNotebookInput := &sagemaker.DescribeCodeRepositoryInput{
+		CodeRepositoryName: aws.String(d.Id()),
+	}
+	CodeRepository, err := conn.DescribeCodeRepository(describeNotebookInput)
```
```suggestion
	input := &sagemaker.DescribeCodeRepositoryInput{
		CodeRepositoryName: aws.String(d.Id()),
	}

	codeRepository, err := conn.DescribeCodeRepository(describeNotebookInput)
```

and change `CodeRepository` to `codeRepository` below.

ewbankkit replied:
Also, maybe move to an internal `finder` package (and reuse that function in `testAccCheckAWSSagemakerCodeRepositoryDestroy` and `testAccCheckAWSSagemakerCodeRepositoryExists`)?

DrFaust92 replied:
added

ewbankkit commented on aws/resource_aws_sagemaker_code_repository.go:
```diff
@@ -0,0 +1,205 @@
+	log.Printf("[DEBUG] sagemaker code repository create config: %#v", *createOpts)
+	_, err := conn.CreateCodeRepository(createOpts)
+	if err != nil {
+		return fmt.Errorf("Error creating Sagemaker code repository: %w", err)
```
```suggestion
		return fmt.Errorf("Error creating SageMaker code repository: %w", err)
```

ewbankkit commented on aws/resource_aws_sagemaker_code_repository.go:
```diff
@@ -0,0 +1,205 @@
+
+	}
+
+	if err := d.Set("code_repository_name", CodeRepository.CodeRepositoryName); err != nil {
+		return fmt.Errorf("error setting code_repository_name for sagemaker code repository (%s): %w", d.Id(), err)
+	}
+
+	if err := d.Set("arn", CodeRepository.CodeRepositoryArn); err != nil {
+		return fmt.Errorf("error setting arn for sagemaker code repository (%s): %w", d.Id(), err)
+	}
```
No need to check `err` for simple attribute values.

DrFaust92 replied:
done

ewbankkit commented on aws/resource_aws_sagemaker_code_repository.go:
```diff
@@ -0,0 +1,205 @@
+	log.Printf("[DEBUG] sagemaker code repository update config: %#v", *input)
+	_, err := conn.UpdateCodeRepository(input)
+	if err != nil {
+		return fmt.Errorf("Error updating Sagemaker code repository: %w", err)
```
```suggestion
		return fmt.Errorf("error updating SageMaker code repository: %w", err)
```

ewbankkit commented on aws/resource_aws_sagemaker_code_repository.go:
```diff
@@ -0,0 +1,205 @@
+			log.Printf("[WARN] Unable to find SageMaker code repository (%s); removing from state", d.Id())
+			return nil
+		}
+		return fmt.Errorf("error finding sagemaker code repository (%s): %w", d.Id(), err)
```
```suggestion
		return fmt.Errorf("error reading SageMaker code repository (%s): %w", d.Id(), err)
```

ewbankkit commented on aws/resource_aws_sagemaker_code_repository.go:
```diff
@@ -0,0 +1,205 @@
+		if isAWSErr(err, "ValidationException", "Cannot find CodeRepository") {
+			return nil
+		}
+		return fmt.Errorf("error trying to delete sagemaker code repository (%s): %w", d.Id(), err)
```
```suggestion
		return fmt.Errorf("error deleting SageMaker code repository (%s): %w", d.Id(), err)
```

ewbankkit commented:
Verified acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSSagemakerCodeRepository_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSSagemakerCodeRepository_ -timeout 120m
=== RUN   TestAccAWSSagemakerCodeRepository_basic
=== PAUSE TestAccAWSSagemakerCodeRepository_basic
=== RUN   TestAccAWSSagemakerCodeRepository_gitConfig_branch
=== PAUSE TestAccAWSSagemakerCodeRepository_gitConfig_branch
=== RUN   TestAccAWSSagemakerCodeRepository_gitConfig_secret
=== PAUSE TestAccAWSSagemakerCodeRepository_gitConfig_secret
=== RUN   TestAccAWSSagemakerCodeRepository_disappears
=== PAUSE TestAccAWSSagemakerCodeRepository_disappears
=== CONT  TestAccAWSSagemakerCodeRepository_basic
=== CONT  TestAccAWSSagemakerCodeRepository_disappears
=== CONT  TestAccAWSSagemakerCodeRepository_gitConfig_secret
=== CONT  TestAccAWSSagemakerCodeRepository_gitConfig_branch
--- PASS: TestAccAWSSagemakerCodeRepository_disappears (17.27s)
--- PASS: TestAccAWSSagemakerCodeRepository_basic (23.32s)
--- PASS: TestAccAWSSagemakerCodeRepository_gitConfig_branch (29.28s)
--- PASS: TestAccAWSSagemakerCodeRepository_gitConfig_secret (46.38s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	46.431s
```

DrFaust92 commented:
addressed PR comments:

```
--- PASS: TestAccAWSSagemakerCodeRepository_disappears (26.96s)
--- PASS: TestAccAWSSagemakerCodeRepository_basic (36.97s)
--- PASS: TestAccAWSSagemakerCodeRepository_gitConfig_branch (37.72s)
--- PASS: TestAccAWSSagemakerCodeRepository_gitConfig_secret (85.42s)
```

DrFaust92 commented:
Added test for code repo usage in sagemaker notebook instance:

```
--- PASS: TestAccAWSSagemakerNotebookInstance_default_code_repository_sagemakerRepo (1165.81s)
```

ewbankkit commented on aws/internal/service/sagemaker/finder/finder.go:
```diff
@@ -0,0 +1,25 @@
+	"github.com/aws/aws-sdk-go/service/sagemaker"
+)
+
+// FincCodeRepositoryByName returns the code repository corresponding to the specified name.
+// Returns nil if no code repository is found.
+func FincCodeRepositoryByName(conn *sagemaker.SageMaker, name string) (*sagemaker.DescribeCodeRepositoryOutput, error) {
```
```suggestion
// CodeRepositoryByName returns the code repository corresponding to the specified name.
// Returns nil if no code repository is found.
func CodeRepositoryByName(conn *sagemaker.SageMaker, name string) (*sagemaker.DescribeCodeRepositoryOutput, error) {
```

No need for the `Find` prefix as the package is named `finder`.

ewbankkit reviewed (approved):
LGTM.

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSSagemakerCodeRepository_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSSagemakerCodeRepository_ -timeout 120m
=== RUN   TestAccAWSSagemakerCodeRepository_basic
=== PAUSE TestAccAWSSagemakerCodeRepository_basic
=== RUN   TestAccAWSSagemakerCodeRepository_gitConfig_branch
=== PAUSE TestAccAWSSagemakerCodeRepository_gitConfig_branch
=== RUN   TestAccAWSSagemakerCodeRepository_gitConfig_secret
=== PAUSE TestAccAWSSagemakerCodeRepository_gitConfig_secret
=== RUN   TestAccAWSSagemakerCodeRepository_disappears
=== PAUSE TestAccAWSSagemakerCodeRepository_disappears
=== CONT  TestAccAWSSagemakerCodeRepository_basic
=== CONT  TestAccAWSSagemakerCodeRepository_disappears
=== CONT  TestAccAWSSagemakerCodeRepository_gitConfig_secret
=== CONT  TestAccAWSSagemakerCodeRepository_gitConfig_branch
--- PASS: TestAccAWSSagemakerCodeRepository_disappears (17.00s)
--- PASS: TestAccAWSSagemakerCodeRepository_gitConfig_branch (23.74s)
--- PASS: TestAccAWSSagemakerCodeRepository_basic (23.86s)
--- PASS: TestAccAWSSagemakerCodeRepository_gitConfig_secret (47.05s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	47.118s
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSSagemakerNotebookInstance_default_code_repository_sagemakerRepo\|TestAccAWSSagemakerNotebookInstance_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSSagemakerNotebookInstance_default_code_repository_sagemakerRepo\|TestAccAWSSagemakerNotebookInstance_basic -timeout 120m
=== RUN   TestAccAWSSagemakerNotebookInstance_basic
=== PAUSE TestAccAWSSagemakerNotebookInstance_basic
=== RUN   TestAccAWSSagemakerNotebookInstance_default_code_repository_sagemakerRepo
=== PAUSE TestAccAWSSagemakerNotebookInstance_default_code_repository_sagemakerRepo
=== CONT  TestAccAWSSagemakerNotebookInstance_basic
=== CONT  TestAccAWSSagemakerNotebookInstance_default_code_repository_sagemakerRepo
--- PASS: TestAccAWSSagemakerNotebookInstance_basic (336.17s)
--- PASS: TestAccAWSSagemakerNotebookInstance_default_code_repository_sagemakerRepo (983.18s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	983.242s
```

DrFaust92 commented:
Rebased

ewbankkit commented:
Verified acceptance tests after rebase:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSSagemakerCodeRepository_\|TestAccAWSSagemakerNotebookInstance_default_code_repository_sagemakerRepo\|TestAccAWSSagemakerNotebookInstance_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSSagemakerCodeRepository_\|TestAccAWSSagemakerNotebookInstance_default_code_repository_sagemakerRepo\|TestAccAWSSagemakerNotebookInstance_basic -timeout 120m
=== RUN   TestAccAWSSagemakerCodeRepository_basic
=== PAUSE TestAccAWSSagemakerCodeRepository_basic
=== RUN   TestAccAWSSagemakerCodeRepository_gitConfig_branch
=== PAUSE TestAccAWSSagemakerCodeRepository_gitConfig_branch
=== RUN   TestAccAWSSagemakerCodeRepository_gitConfig_secret
=== PAUSE TestAccAWSSagemakerCodeRepository_gitConfig_secret
=== RUN   TestAccAWSSagemakerCodeRepository_disappears
=== PAUSE TestAccAWSSagemakerCodeRepository_disappears
=== RUN   TestAccAWSSagemakerNotebookInstance_basic
=== PAUSE TestAccAWSSagemakerNotebookInstance_basic
=== RUN   TestAccAWSSagemakerNotebookInstance_default_code_repository_sagemakerRepo
=== PAUSE TestAccAWSSagemakerNotebookInstance_default_code_repository_sagemakerRepo
=== CONT  TestAccAWSSagemakerCodeRepository_basic
=== CONT  TestAccAWSSagemakerNotebookInstance_default_code_repository_sagemakerRepo
=== CONT  TestAccAWSSagemakerNotebookInstance_basic
=== CONT  TestAccAWSSagemakerCodeRepository_gitConfig_secret
=== CONT  TestAccAWSSagemakerCodeRepository_disappears
=== CONT  TestAccAWSSagemakerCodeRepository_gitConfig_branch
2020/10/29 14:34:13 [WARN] Truncating attribute path of 0 diagnostics for TypeSet
2020/10/29 14:34:13 [WARN] Truncating attribute path of 0 diagnostics for TypeSet
2020/10/29 14:34:13 [WARN] Truncating attribute path of 0 diagnostics for TypeSet
--- PASS: TestAccAWSSagemakerCodeRepository_disappears (19.23s)
--- PASS: TestAccAWSSagemakerCodeRepository_gitConfig_branch (24.93s)
--- PASS: TestAccAWSSagemakerCodeRepository_basic (25.02s)
--- PASS: TestAccAWSSagemakerCodeRepository_gitConfig_secret (50.26s)
--- PASS: TestAccAWSSagemakerNotebookInstance_basic (393.80s)
--- PASS: TestAccAWSSagemakerNotebookInstance_default_code_repository_sagemakerRepo (1123.79s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	1123.859s
```

breathingdust commented:
LGTM :rocket: Thanks @DrFaust92 

Verified Acceptance Tests in Commercial (us-west-2)

```console
make testacc TESTARGS='-run=TestAccAWSSagemakerCodeRepository_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSSagemakerCodeRepository_ -timeout 120m
=== RUN   TestAccAWSSagemakerCodeRepository_basic
=== PAUSE TestAccAWSSagemakerCodeRepository_basic
=== RUN   TestAccAWSSagemakerCodeRepository_gitConfig_branch
=== PAUSE TestAccAWSSagemakerCodeRepository_gitConfig_branch
=== RUN   TestAccAWSSagemakerCodeRepository_gitConfig_secret
=== PAUSE TestAccAWSSagemakerCodeRepository_gitConfig_secret
=== RUN   TestAccAWSSagemakerCodeRepository_disappears
=== PAUSE TestAccAWSSagemakerCodeRepository_disappears
=== CONT  TestAccAWSSagemakerCodeRepository_basic
=== CONT  TestAccAWSSagemakerCodeRepository_disappears
=== CONT  TestAccAWSSagemakerCodeRepository_gitConfig_secret
=== CONT  TestAccAWSSagemakerCodeRepository_gitConfig_branch
--- PASS: TestAccAWSSagemakerCodeRepository_disappears (11.08s)
--- PASS: TestAccAWSSagemakerCodeRepository_gitConfig_branch (15.07s)
--- PASS: TestAccAWSSagemakerCodeRepository_basic (15.27s)
--- PASS: TestAccAWSSagemakerCodeRepository_gitConfig_secret (37.54s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	39.071s
```

Note that Code Repositories are not currently supported in the GovCloud region.
